File: pkg/material/checker.go
package material

import (
	"math"
	"math/rand"

	"sif/gogs/eric/relray/pkg/geometry"
	"sif/gogs/eric/relray/pkg/spectrum"
	"sif/gogs/eric/relray/pkg/vec"
)

// Checker is a Lambertian diffuse material with a checkerboard pattern
// alternating between two spectral reflectances. The pattern is computed
// in the tangent plane of the surface (perpendicular to the hit normal),
// so it produces a proper checkerboard on every flat face of any shape.
type Checker struct {
	Even, Odd spectrum.SPD
	Scale     float64 // size of each checker square in world units
}

func (c *Checker) Scatter(inDir vec.Vec3, hit geometry.Hit, rng *rand.Rand) ScatterResult {
	scattered := hit.Normal.Add(RandomUnitVec(rng))
	if scattered.LengthSq() < 1e-12 {
		scattered = hit.Normal
	}
	return ScatterResult{
		Scattered:   true,
		OutDir:      scattered.Normalize(),
		Reflectance: c.reflectanceAt(hit.Point, hit.Normal),
	}
}

func (c *Checker) Emitted(hit geometry.Hit) spectrum.SPD {
	return spectrum.SPD{}
}

func (c *Checker) reflectanceAt(p, n vec.Vec3) spectrum.SPD {
	// Build a tangent frame from the surface normal.
	ref := vec.Vec3{X: 1, Y: 0, Z: 0}
	if math.Abs(n.X) > 0.9 {
		ref = vec.Vec3{X: 0, Y: 1, Z: 0}
	}
	t1 := n.Cross(ref).Normalize()
	t2 := n.Cross(t1)

	inv := 1.0 / c.Scale
	iu := int(math.Floor(p.Dot(t1) * inv))
	iv := int(math.Floor(p.Dot(t2) * inv))
	if (iu+iv)%2 == 0 {
		return c.Even
	}
	return c.Odd
}

File: rust/src/material/checker.rs
use crate::geometry::Hit;
use crate::spectrum::Spd;
use crate::vec::Vec3;
use rand::rngs::SmallRng;

use super::{random_unit_vec, Material, ScatterResult};

/// Lambertian diffuse material with a checkerboard pattern alternating
/// between two spectral reflectances. The pattern is computed in the
/// tangent plane of the surface (perpendicular to the hit normal), so it
/// produces a proper checkerboard on every flat face of any shape.
pub struct Checker {
    pub even: Spd,
    pub odd: Spd,
    /// Size of each checker square in world units.
    pub scale: f64,
}

impl Checker {
    fn reflectance_at(&self, p: Vec3, n: Vec3) -> Spd {
        // Build a tangent frame from the surface normal.
        let reference = if n.x.abs() > 0.9 {
            Vec3 { x: 0.0, y: 1.0, z: 0.0 }
        } else {
            Vec3 { x: 1.0, y: 0.0, z: 0.0 }
        };
        let t1 = n.cross(reference).normalize();
        let t2 = n.cross(t1);

        let inv = 1.0 / self.scale;
        let iu = (p.dot(t1) * inv).floor() as i64;
        let iv = (p.dot(t2) * inv).floor() as i64;
        if (iu + iv) % 2 == 0 {
            self.even
        } else {
            self.odd
        }
    }
}

impl Material for Checker {
    fn scatter(&self, _in_dir: Vec3, hit: &Hit, rng: &mut SmallRng) -> ScatterResult {
        let mut scattered = hit.normal + random_unit_vec(rng);
        if scattered.length_sq() < 1e-12 {
            scattered = hit.normal;
        }
        ScatterResult {
            scattered: true,
            out_dir: scattered.normalize(),
            reflectance: self.reflectance_at(hit.point, hit.normal),
        }
    }

    fn emitted(&self, _hit: &Hit) -> Spd {
        Spd::default()
    }
}
